### PR TITLE
wit/bindgen: Move wasmimports/wasmexports to a dedicated file

### PR DESCRIPTION
**Pull request by lxfontes**
Move `wasmimport` and `wasmexport` declarations to a dedicated file (named `<package>.wasm.go` ).

Rationale: The current `.wit.go` file carries Go and functions intended to be used as wasm bridge ( wasm import / export ). Separating go code from wasm bridge allows readers to identify what is crossing the wasm boundary and gate them if needed. A secondary benefit is paving the way for `wit-bindgen-go` plugins that utilize the existing ABI and bridge into other implementations. The 2 that come to mind:

1. test/mocking for component unit testing ( with & without wasmtime )
2. networked protocols ( ex https://github.com/wrpc/wrpc )

Sample codegen for wasi:logging

logging.wit.go
```
//go:nosplit
func Log(level Level, context string, message string) {
	level0 := (uint32)(level)
	context0, context1 := cm.LowerString(context)
	message0, message1 := cm.LowerString(message)
	wasmimport_Log((uint32)(level0), (*uint8)(context0), (uint32)(context1), (*uint8)(message0), (uint32)(message1))
	return
}
```

logging.wasm.go
```
// Code generated by wit-bindgen-go. DO NOT EDIT.

package logging

// This file contains all wasmimport/wasmexport functions for "wasi:logging/logging".

//go:wasmimport wasi:logging/logging log
//go:noescape
func wasmimport_Log(level0 uint32, context0 *uint8, context1 uint32, message0 *uint8, message1 uint32)
```

Edit: a future PR will implement build tags for the `<package>.wasm.go` file, e.g. `//go:build wasm || wasm32 || tinygo.wasm`.

### CODE REVIEW

File: cmd/wit-bindgen-go/cmd/generate/generate.go [hunk 1]
```diff
@@ -52,6 +52,13 @@ var Command = &cli.Command{
 			Config:   cli.StringConfig{TrimSpace: true},
 			Usage:    "Import path for the Component Model utility package, e.g. github.com/bytecodealliance/wasm-tools-go/cm",
 		},
+		&cli.StringFlag{
+			Name:     "extern-tag",
+			Value:    "",
+			OnlyOnce: true,
+			Config:   cli.StringConfig{TrimSpace: true},
+			Usage:    "Build tag for wasmimport/wasmexport functions. See https://pkg.go.dev/go/build",
+		},
```
Comment (ydnar) [inline cmd/wit-bindgen-go/cmd/generate/generate.go:61]: Let’s omit this. The build tags can be simple now: `//go:build wasm || wasm32 || tinygo.wasm`.

File: wit/bindgen/generator.go [hunk 2]
```diff
@@ -1800,31 +1800,33 @@ func (g *generator) defineImportedFunction(_ wit.Ident, f *wit.Function, decl fu
 	b.WriteString("}\n\n")
 
 	// Emit wasmimport function
-	stringio.Write(&b, "//go:wasmimport ", decl.linkerName, "\n")
-	b.WriteString("//go:noescape\n")
-	b.WriteString("func ")
+	externFile := g.externFileFor(owner)
```
Comment (ydnar) [inline wit/bindgen/generator.go:1803]: Let’s call this `wasmFileFor`.

File: wit/bindgen/generator.go [hunk 3]
```diff
@@ -1800,31 +1800,33 @@ func (g *generator) defineImportedFunction(_ wit.Ident, f *wit.Function, decl fu
...
 	}
 
 	if t, ok := compoundResults.typ.(*wit.TypeDef); ok {
 		td, _ := g.typeDecl(dir, t)
-		stringio.Write(&b, "// ", td.name, " represents the flattened function results for [", decl.wasm.name, "].\n")
-		stringio.Write(&b, "// See the Canonical ABI flattening rules for more information.\n")
-		stringio.Write(&b, "type ", td.name, " ", g.typeDefRep(file, dir, t, td.name), "\n\n")
+		stringio.Write(externFile, "// ", td.name, " represents the flattened function results for [", decl.wasm.name, "].\n")
+		stringio.Write(externFile, "// See the Canonical ABI flattening rules for more information.\n")
+		stringio.Write(externFile, "type ", td.name, " ", g.typeDefRep(externFile, dir, t, td.name), "\n\n")
```
Comment (ydnar) [inline wit/bindgen/generator.go:1829]: This type should be emitted in the main file, not the wasm file, so implementors don’t need to reimplement this type when compiling without `wasm` tag.

File: wit/bindgen/generator.go [hunk 4]
```diff
@@ -2171,6 +2175,22 @@ func (g *generator) exportsFileFor(id wit.Ident) *gen.File {
 	return file
 }
 
+func (g *generator) externFileFor(id wit.Ident) *gen.File {
```
Comment (ydnar) [inline wit/bindgen/generator.go:2178]: `wasmFileFor`

File: wit/bindgen/generator.go [hunk 5]
```diff
@@ -2171,6 +2175,22 @@ func (g *generator) exportsFileFor(id wit.Ident) *gen.File {
...
+	pkg := g.packageFor(id)
+	file := pkg.File(id.Extension + ".extern.go")
+	file.GeneratedBy = g.opts.generatedBy
+	if len(file.Header) == 0 {
+		var b strings.Builder
+		if g.opts.externTag != "" {
+			stringio.Write(&b, "//go:build ", g.opts.externTag, "\n")
+		}
+		stringio.Write(&b, "// This file contains all wasmimport/wasmexport functions for \"", id.String(), "\".\n\n")
```
Comment (ydnar) [inline wit/bindgen/generator.go:2187]:
```suggestion
		stringio.Write(&b, "// This file contains wasmimport and wasmexport declarations for \"", id.String(), "\".\n\n")
```

File: wit/bindgen/generator.go [hunk 6]
```diff
@@ -2171,6 +2175,22 @@ func (g *generator) exportsFileFor(id wit.Ident) *gen.File {
...
 
+func (g *generator) externFileFor(id wit.Ident) *gen.File {
+	pkg := g.packageFor(id)
+	file := pkg.File(id.Extension + ".extern.go")
+	file.GeneratedBy = g.opts.generatedBy
+	if len(file.Header) == 0 {
+		var b strings.Builder
+		if g.opts.externTag != "" {
+			stringio.Write(&b, "//go:build ", g.opts.externTag, "\n")
+		}
```
Comment (ydnar) [inline wit/bindgen/generator.go:2186]: Let’s omit this configuration knob / CLI arg unless it’s absolutely necessary.

File: wit/bindgen/generator.go [hunk 7]
```diff
@@ -2171,6 +2175,22 @@ func (g *generator) exportsFileFor(id wit.Ident) *gen.File {
...
+		stringio.Write(&b, "// This file contains all wasmimport/wasmexport functions for \"", id.String(), "\".\n\n")
+		file.Header = b.String()
+	}
+	file.Trailer = "\n"
```
Comment (ydnar) [inline wit/bindgen/generator.go:2190]: This should be unnecessary.

**Review by ydnar — CHANGES_REQUESTED** [review]
This is great, thanks for drafting the PR. I support this direction.

**Comment by ydnar** [comment]
FYI, this project follows the Go project conventions for PR names and commit messages:

```
package: present tense verb describing change

More context...
```

File: wit/bindgen/generator.go [hunk 8]
```diff
@@ -39,6 +39,9 @@ const (
 
 	// Define Go GC shape types for variant and result storage.
 	experimentCreateShapeTypes = true
+
+	// build tag for '.wasm.go' files
+	wasmBuildTag = "wasm || wasm32 || tinygo.wasm"
```
Comment (ydnar) [inline wit/bindgen/generator.go:44]: This could probably be moved up beneath const `cmPackage` above.

**Review by ydnar — APPROVED** [review]
One minor change, we could do now or later.

**Comment by ydnar** [comment]
Do you want to update the PR description to reflect current changes?

**Comment by lxfontes** [comment]
> I checked out the branch and ran `go test ./wit/bindgen -write`, which writes all the generated Go to the filesystem. Afterward, I ran `go test ./...`, which fails because the `wasmimport` functions were no longer found.
> 
> I think it’s important that `go test ./...` on Go packages generated by `wit/bindgen` pass, or at the very least, _do not fail_.
> 
> If the wasm functions are sequestered in a .wasm.go file, how do you think we can accomplish this?

It's the `//go:build` gate on that file.
That's where the `wit-bindgen-go --wasmBuildTag` was helpful, but could also cause confusion.

I do agree that failing `go test` is a blocker and think we have 2 options:
- don't add the `//go:build` part to this PR
- duplicate wasm file, with a build tag negating the wasm file's tags

Duplicating the wasm file doesn't seem to add any value comparing to removing the build tag.
The follow-up PR I was planning is a mock generator for non-wasm builds, performing the lift/lowering into existing structs. That might be a better time to add the `//go:build` gate to wasm files.

Still have the build tag in the PR and can remove it if that makes sense 👍

File: wit/bindgen/testdata_test.go [hunk 9]
```diff
@@ -116,11 +116,10 @@ func validateGeneratedGo(t *testing.T, res *wit.Resolve, origin string) {
 	pkgMap := make(map[string]*gen.Package)
 
 	cfg := &packages.Config{
-		Mode:       packages.NeedName | packages.NeedFiles | packages.NeedCompiledGoFiles | packages.NeedImports | packages.NeedTypes | packages.NeedTypesInfo | packages.NeedTypesSizes,
-		Dir:        out,
-		Fset:       token.NewFileSet(),
-		Overlay:    make(map[string][]byte),
-		BuildFlags: []string{"-tags=wasm"},
```
Comment (ydnar) [inline wit/bindgen/testdata_test.go]: Hah, nice.

**Review by ydnar — APPROVED** [review]

**Comment by ydnar** [comment]
I should have rebased on `main` before merging, as the type alias PR uncovered a subtle bug in the code that generates Go representation of variant and result types.

**Pull request merged**